Add user guide for namespaced queries (#1191)
block/elasticgraph: config/site/src/guides/namespaced-queries.md

fields ElasticGraph generates for an indexed type are added directly to `Query`. For example, an indexed `Widget` type 
produces `Query.widgets` and `Query.widgetAggregations`.

When ElasticGraph is composed into a federated supergraph alongside other subgraphs, `Query`
can become crowded and fields from different subgraphs can collide. A _namespace type_
lets you group ElasticGraph's root query fields under a nested path—for example, `Query.olap.widgets`
instead of `Query.widgets`. This improves discoverability and isolation of domain specific types.

## Minimal Example

A namespace is an object type declared with [`namespace_type`](/elasticgraph/api-docs/main/ElasticGraph/SchemaDefinition/API.html#namespace_type-instance_method).
You can route an indexed type's root fields to the namespace by passing `on:` to [`root_query_fields`](/elasticgraph/api-docs/main/ElasticGraph/SchemaDefinition/Mixins/HasIndices.html#root_query_fields-instance_method),
then expose the namespace type as a field on `Query`.

{% include copyable_code_snippet.html language="ruby" code='ElasticGraph.define_schema do |schema|
  schema.namespace_type "OlapQuery"

  schema.on_root_query_type do |t|

>     t.field "olap", "OlapQuery!"

Reviewer: ```suggestion
    t.field "olap", "OlapQuery"
```

As per [our docs](https://github.com/block/elasticgraph/blob/525d1ba9aba46f36d2c5de2a9307d44e3636a3a1/elasticgraph-schema_definition/lib/elastic_graph/schema_definition/schema_elements/type_with_subfields.rb#L155-L161) we generally discourage the use of `!` on fields besides cases such as lists (I recommend `[ElementType!]!` to guard against `null` values inside and outside the list, and the list already provides a "no data" representation in the form of `[]` so there's always a way for the GraphQL endpoint to return no data even though it's non-nullable).

  end

  schema.object_type "Widget" do |t|
    t.field "id", "ID"
    t.field "name", "String"
    t.index "widgets"
    t.root_query_fields plural: "widgets", on: "OlapQuery"
  end

> end' %}

Reviewer: One of our [guiding principles](https://block.github.io/elasticgraph/guides/guiding-principles/) is:

> **We aim to validate all snippets and code examples at this website.**
>
> Our documentation is so much more useful if users can trust that the code snippets we provide always work. When a change to the codebase breaks an example from the website, we want to be notified about it so that we can fix it. The CI build includes full validation of website code snippets.

To support that, our website has a system where we can define snippets like this in the context of an example project which gets validated as part of the CI build, and then we can pull the snippet in from there instead of it being written by hand.

I asked claude to write up how this system works; below is what it came up with. Can you migrate this snippet (and your snippets below) into that system?

---------

### 1. Example projects

We have [example projects](https://github.com/block/elasticgraph/tree/525d1ba9aba46f36d2c5de2a9307d44e3636a3a1/config/site/examples) (`config/site/examples/`) that contain real, working ElasticGraph schemas and queries. Currently: `music`, `music_simplified`, and `custom_resolver`.

### 2. Snippet markers in source files

Within those example projects, we fence reusable snippets with `:snippet-start:` / `:snippet-end:` markers. For example, in [`config/site/examples/custom_resolver/schema.rb`](https://github.com/block/elasticgraph/blob/525d1ba9aba46f36d2c5de2a9307d44e3636a3a1/config/site/examples/custom_resolver/schema.rb#L14-L20):

```ruby
# :snippet-start: register_graphql_resolver
require(require_path = "roll_dice_resolver")
schema.register_graphql_resolver :roll_dice,
  RollDiceResolver,
  defined_at: require_path,
  number_of_dice: 2
# :snippet-end:
```

GraphQL queries are just standalone `.graphql` files under a `queries/` directory, organized by category (e.g. [`config/site/examples/music/queries/filtering/FindArtistsFormedIn90s.graphql`](https://github.com/block/elasticgraph/blob/525d1ba9aba46f36d2c5de2a9307d44e3636a3a1/config/site/examples/music/queries/filtering/FindArtistsFormedIn90s.graphql)).

### 3. Extraction into Jekyll data files

The [`site:examples:extract_snippets` Rake task](https://github.com/block/elasticgraph/blob/525d1ba9aba46f36d2c5de2a9307d44e3636a3a1/config/site/Rakefile#L440-L494) extracts these into YAML data files under `config/site/src/_data/`. It produces entries like `custom_resolver.yaml` (with `files` and `snippets` sections) and `music_queries.yaml` (queries organized by category).

The extraction logic is in [`extract_fenced_snippets_from_file`](https://github.com/block/elasticgraph/blob/525d1ba9aba46f36d2c5de2a9307d44e3636a3a1/config/site/Rakefile#L515-L536).

### 4. Including snippets in documentation

Documentation pages reference extracted snippets via a Jekyll include using a dot-separated data path. From [`_includes/copyable_code_snippet.html`](https://github.com/block/elasticgraph/blob/525d1ba9aba46f36d2c5de2a9307d44e3636a3a1/config/site/src/_includes/copyable_code_snippet.html#L11-L19):

```liquid
{% include copyable_code_snippet.html language="graphql" data="music_queries.filtering.FindArtistsFormedIn90s" %}
```

The `data` path is resolved by walking `site.data` (the YAML files), e.g. `music_queries` -> `filtering` -> `FindArtistsFormedIn90s`.

You can see many examples of this in [`guides/custom-graphql-resolvers.md`](https://github.com/block/elasticgraph/blob/525d1ba9aba46f36d2c5de2a9307d44e3636a3a1/config/site/src/guides/custom-graphql-resolvers.md).

### 5. Validation in CI

The build enforces correctness at multiple levels:

- **Schema/query validation**: Each example's `extract_snippets` task depends on `validate` ([line 461](https://github.com/block/elasticgraph/blob/525d1ba9aba46f36d2c5de2a9307d44e3636a3a1/config/site/Rakefile#L461)), which dumps schema artifacts and runs any custom `validate.rb`. Queries are validated against the schema via `query_registry:validate_queries` ([line 480](https://github.com/block/elasticgraph/blob/525d1ba9aba46f36d2c5de2a9307d44e3636a3a1/config/site/Rakefile#L480)).
- **Empty snippet detection**: After Jekyll builds, [`validate_no_empty_code_snippets`](https://github.com/block/elasticgraph/blob/525d1ba9aba46f36d2c5de2a9307d44e3636a3a1/config/site/support/validate_no_empty_code_snippets.rb) parses the HTML and fails the build if any `copyable_code_snippet` include resolved to empty content (meaning the `data=` path was invalid).

So if a code change breaks an example schema or query, CI catches it.

The namespace type is named `OlapQuery` (GraphQL types are PascalCase by convention) and is exposed

> as the `olap` field on `Query` (camelCase). Throughout this guide, "namespace type" refers to the

Reviewer: ```suggestion
The namespace type is named `OlapQuery` and is exposed
as the `olap` field on `Query`. Throughout this guide, "namespace type" refers to the
```

I don't think we need commentary on the casing here.

> type itself and field references like `Query.olap` refer to the field on `Query` that returns it.

Reviewer: > Throughout this guide, "namespace type" refers to the
> type itself and field references like `Query.olap` refer to the field on `Query` that returns it.

This reads like AI slop. This seems to be stating things that are self-evident to any reasonably intelligent human readable.

This produces a GraphQL API where `Widgets` are queried through `olap`:

{% include copyable_code_snippet.html language="graphql" code="query {
  olap {
    widgets(first: 10) {
      nodes { id name }
    }
    widgetAggregations {
      nodes { count }
    }
  }
}" %}

You don't need to wire up a resolver for `Query.olap`. ElasticGraph auto-resolves any no-argument
field whose return type is a namespace type with an inert passthrough object, so each child field's

> own resolver runs against it.

Reviewer: ```suggestion
You don't need to wire up a resolver for `Query.olap`. ElasticGraph auto-resolves any no-argument
field whose return type is a namespace type.
```

I don't think "inert passthrough object" is something that end-users need to understand--just knowing it auto-resolves (without knowing the details) should be sufficient.

## Nested Example

Namespace types can be nested inside other namespace types. The same auto-resolution applies, so
you don't have to configure a resolver for any intermediate field.

{% include copyable_code_snippet.html language="ruby" code='ElasticGraph.define_schema do |schema|
  schema.namespace_type "OlapQuery" do |t|

>     t.field "domain", "DomainQuery!"

Reviewer: ```suggestion
    t.field "domain", "DomainQuery"
```

  end

  schema.namespace_type "DomainQuery"

  schema.on_root_query_type do |t|

>     t.field "olap", "OlapQuery!"

Reviewer: ```suggestion
    t.field "olap", "OlapQuery"
```

  end

  schema.object_type "Widget" do |t|
    t.field "id", "ID"
    t.index "widgets"
    t.root_query_fields plural: "widgets", on: "DomainQuery"
  end
end' %}

Widgets are now queried at `Query.olap.domain.widgets`.

## Tradeoffs

### Single Target per Indexed Type

An indexed type's root fields (`plural` and `singular`) are always placed together on one target
type; either `Query` (the default) or a single namespace type. You cannot split them; for example,
you cannot put `widgets` on `Query` and `widgetAggregations` on `OlapQuery`. If you need different
groupings for the list field and the aggregation field, consider whether the namespace type is

> actually the right grouping, or model the split at the supergraph level.

Reviewer: ```suggestion
```

I don't think there's much signal here among the AI noise. It doesn't really describe a tradeoff in a meaningful sense. The examples above have already shown that `widgets` and `widgetAggregations` both got defined on the namespace type.

## Apollo Federation

If you expose ElasticGraph through [elasticgraph-apollo](/elasticgraph/api-docs/{{ site.data.doc_versions.latest_version }}/ElasticGraph/Apollo.html),
namespace types appear in the `_service` SDL like any other type. Depending on your composition
strategy, you may need to apply federation directives:

- `@shareable` — if another subgraph also defines a type with the same name and overlapping fields,
  use `apollo_shareable` on the namespace type (and on shared fields) so Apollo composition allows
  the overlap.
- `@inaccessible` — use `apollo_inaccessible` on fields you don't want exposed in the final
  supergraph schema.

Namespace types you declare behave like any user-defined type — they are not tagged with
`@shareable` automatically, even if you've called `tag_built_in_types_with`. Apply the directive

> explicitly when your supergraph composition requires it.

Reviewer: ```suggestion
```

Similar to the above, I don't think there's much signal here. I don't think we need to explain "Namespace types you declare behave like any user-defined type" or explain `@shareable`/`@inaccessible`--those things aren't special to namespace types. If we want to cover this stuff in a guide, I'd rather it be on a brand-new guide page that's about the Apollo federation support as a whole.
